exportlib.php

```diff
@@ -84,47 +84,41 @@ public function create_file($courseid, $feedbackid) {
 }
 
 /**
- * @author Felix Di Lenarda
- * rankings CSV export class
+ * Export feedback data for an entire feedbacj.
+ *
+ * @package block
+ * @subpackage coursefeedback
+ * @copyright innoCampus, TU Berlin
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
-class rankingexport {
+class ranking_exporter {
     protected $feedback = null;
     protected $questionid = null;
-    protected $seperator = ";";
-
-    private $content = "";
+    protected $csvexportwriter;
 
     public function __construct($feedback = null, $question = null) {
-        global $DB;
-
-        if ($fb = $DB->get_record("block_coursefeedback", array("id" => $feedback))) {
-            $this->feedback = $fb;
-            $this->questionid = $question;
-        } else {
-            print_error("feedbacknotfound", "error");
-            exit(0);
-        }
+        $this->csvexportwriter = new csv_export_writer();
     }
-
-    public function export() {
+    public function create_file($feedbackid, $questionid = 0 )
+    {
         global $DB;
-        $writer = new csv_export_writer('semicolon');
-        $writer->set_filename( clean_param(get_string("download_html_filename", "block_coursefeedback"),
-            PARAM_FILE));
+        $this->csvexportwriter->set_filename(get_string("download_html_filename", "block_coursefeedback")
+            . date("_Y-m-d-H-i"));
 
-        $writer->add_data([
-            'Feedbackid: ' . $this->feedback->id,
-            'Feedbackid: ' . $this->feedback->name
+        $feedback = $DB->get_record("block_coursefeedback", ["id" => $feedbackid]);
+        $this->csvexportwriter->add_data([
+            'Feedbackid: ' . $feedback->id,
+            'Feedbackname: ' . $feedback->name
         ]);
 
         // Get questions
-        $qus = block_coursefeedback_get_questions_by_language($this->feedback->id, [current_language()]);
+        $qus = block_coursefeedback_get_questions_by_language($feedback->id, [current_language()]);
         $questions = null;
 
-        if ($this->questionid) {
+        if ($questionid != 0) {
             // Only display one question.
             foreach ($qus as $qu) {
-                if ($qu->questionid == $this->questionid) {
+                if ($qu->questionid == $questionid) {
                     $questions = array($qu);
```

Comment:
Hiernach fehlt m.E. ein `break`, um die Schleife zu unterbrechen. Ich gehe davon aus, dass die Elemente, die von `block_coursefeedback_get_questions_by_language` zurückgegeben werden, einzigartige `questionid` Attribute haben.

Man könnte aber auch den ganzen `if`-Block ersetzen, wenn man [`array_filter`](https://www.php.net/manual/en/function.array-filter.php) verwendet:

```php
$questions = block_coursefeedback_get_questions_by_language($feedback->id, [current_language()]);

if ($questionid != 0) {
    $questions = array_filter($questions, fn($q) => $q->questionid == $questionid);
}
```

Arrow-Funktionen gibt es erst ab PHP 7.4, ansonsten:

```php
...
    $questions = array_filter($questions, function ($q) { return $q->questionid == $questionid; });
```

Reply:
Hab die "array_filter" Variante gewählt und noch "use ($questionid)" ergänzt, sonst hat die Funktion keinen Zugriff...
133d532afd97506c96e5730ab0acca9161b6b022

Reply:
Richtig, das mit dem Scope vergesse ich in PHP immer.
Soweit so gut. Ich hätte wie gesagt, einfach direkt `$questions` verwendet. Also m.E. braucht man die `$qus` Zwischenvariable nicht. Wenn `$questionid == 0` ist, würde man ja das, was `block_coursefeedback_get_questions_by_language` zurückgibt einfach unverändert weiterverwenden. Daher mein Vorschlag mit dem einen `if` ohne den `else` Block. Würde ggü. dem aktuellen Stand noch 3-4 Zeilen sparen, aber wichtig ist es nicht.

Reply:
Richtig :) Ich habs noch geändert:
c71d4c2fcb5ae8a35ce0b6aab67aabcb3bdf6c73